File: src/main/java/com/merge/api/resources/ats/types/ScheduledInterview.java
package com.merge.api.resources.ats.types;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.merge.api.core.ObjectMappers;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonDeserialize(builder = ScheduledInterview.Builder.class)
public final class ScheduledInterview {
    private final Optional<String> id;

    private final Optional<String> remoteId;

    private final Optional<ScheduledInterviewApplication> application;

    private final Optional<ScheduledInterviewJobInterviewStage> jobInterviewStage;

    private final Optional<ScheduledInterviewOrganizer> organizer;

    private final Optional<List<Optional<ScheduledInterviewInterviewersItem>>> interviewers;

    private final Optional<String> location;

    private final Optional<OffsetDateTime> startAt;

    private final Optional<OffsetDateTime> endAt;

    private final Optional<OffsetDateTime> remoteCreatedAt;

    private final Optional<OffsetDateTime> remoteUpdatedAt;

    private final Optional<ScheduledInterviewStatus> status;

    private final Optional<Boolean> remoteWasDeleted;

    private final Optional<OffsetDateTime> createdAt;

    private final Optional<OffsetDateTime> modifiedAt;

    private final Optional<Map<String, JsonNode>> fieldMappings;

    private final Optional<List<RemoteData>> remoteData;

    private ScheduledInterview(
            Optional<String> id,
            Optional<String> remoteId,
            Optional<ScheduledInterviewApplication> application,
            Optional<ScheduledInterviewJobInterviewStage> jobInterviewStage,
            Optional<ScheduledInterviewOrganizer> organizer,
            Optional<List<Optional<ScheduledInterviewInterviewersItem>>> interviewers,
            Optional<String> location,
            Optional<OffsetDateTime> startAt,
            Optional<OffsetDateTime> endAt,
            Optional<OffsetDateTime> remoteCreatedAt,
            Optional<OffsetDateTime> remoteUpdatedAt,
            Optional<ScheduledInterviewStatus> status,
            Optional<Boolean> remoteWasDeleted,
            Optional<OffsetDateTime> createdAt,
            Optional<OffsetDateTime> modifiedAt,
            Optional<Map<String, JsonNode>> fieldMappings,
            Optional<List<RemoteData>> remoteData) {
        this.id = id;
        this.remoteId = remoteId;
        this.application = application;
        this.jobInterviewStage = jobInterviewStage;
        this.organizer = organizer;
        this.interviewers = interviewers;
        this.location = location;
        this.startAt = startAt;
        this.endAt = endAt;
        this.remoteCreatedAt = remoteCreatedAt;
        this.remoteUpdatedAt = remoteUpdatedAt;
        this.status = status;
        this.remoteWasDeleted = remoteWasDeleted;
        this.createdAt = createdAt;
        this.modifiedAt = modifiedAt;
        this.fieldMappings = fieldMappings;
        this.remoteData = remoteData;
    }

    @JsonProperty("id")
    public Optional<String> getId() {
        return id;
    }

    /**
     * @return The third-party API ID of the matching object.
     */
    @JsonProperty("remote_id")
    public Optional<String> getRemoteId() {
        return remoteId;
    }

    /**
     * @return The application being interviewed.
     */
    @JsonProperty("application")
    public Optional<ScheduledInterviewApplication> getApplication() {
        return application;
    }

    /**
     * @return The stage of the interview.
     */
    @JsonProperty("job_interview_stage")
    public Optional<ScheduledInterviewJobInterviewStage> getJobInterviewStage() {
        return jobInterviewStage;
    }

    /**
     * @return The user organizing the interview.
     */
    @JsonProperty("organizer")
    public Optional<ScheduledInterviewOrganizer> getOrganizer() {
        return organizer;
    }

    /**
     * @return Array of <code>RemoteUser</code> IDs.
     */
    @JsonProperty("interviewers")
    public Optional<List<Optional<ScheduledInterviewInterviewersItem>>> getInterviewers() {
        return interviewers;
    }

    /**
     * @return The interview's location.
     */
    @JsonProperty("location")
    public Optional<String> getLocation() {
        return location;
    }

    /**
     * @return When the interview was started.
     */
    @JsonProperty("start_at")
    public Optional<OffsetDateTime> getStartAt() {
        return startAt;
    }

    /**
     * @return When the interview was ended.
     */
    @JsonProperty("end_at")
    public Optional<OffsetDateTime> getEndAt() {
        return endAt;
    }

    /**
     * @return When the third party's interview was created.
     */
    @JsonProperty("remote_created_at")
    public Optional<OffsetDateTime> getRemoteCreatedAt() {
        return remoteCreatedAt;
    }

    /**
     * @return When the third party's interview was updated.
     */
    @JsonProperty("remote_updated_at")
    public Optional<OffsetDateTime> getRemoteUpdatedAt() {
        return remoteUpdatedAt;
    }

    /**
     * @return The interview's status.
     * <ul>
     * <li><code>SCHEDULED</code> - SCHEDULED</li>
     * <li><code>AWAITING_FEEDBACK</code> - AWAITING_FEEDBACK</li>
     * <li><code>COMPLETE</code> - COMPLETE</li>
     * </ul>
     */
    @JsonProperty("status")
    public Optional<ScheduledInterviewStatus> getStatus() {
        return status;
    }

    /**
     * @return Indicates whether or not this object has been deleted in the third party platform.
     */
    @JsonProperty("remote_was_deleted")
    public Optional<Boolean> getRemoteWasDeleted() {
        return remoteWasDeleted;
    }

    @JsonProperty("created_at")
    public Optional<OffsetDateTime> getCreatedAt() {
        return createdAt;
    }

    /**
     * @return This is the datetime that this object was last updated by Merge
     */
    @JsonProperty("modified_at")
    public Optional<OffsetDateTime> getModifiedAt() {
        return modifiedAt;
    }

    @JsonProperty("field_mappings")
    public Optional<Map<String, JsonNode>> getFieldMappings() {
        return fieldMappings;
    }

    @JsonProperty("remote_data")
    public Optional<List<RemoteData>> getRemoteData() {
        return remoteData;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        return other instanceof ScheduledInterview && equalTo((ScheduledInterview) other);
    }

    private boolean equalTo(ScheduledInterview other) {
        return id.equals(other.id)
                && remoteId.equals(other.remoteId)
                && application.equals(other.application)
                && jobInterviewStage.equals(other.jobInterviewStage)
                && organizer.equals(other.organizer)
                && interviewers.equals(other.interviewers)
                && location.equals(other.location)
                && startAt.equals(other.startAt)
                && endAt.equals(other.endAt)
                && remoteCreatedAt.equals(other.remoteCreatedAt)
                && remoteUpdatedAt.equals(other.remoteUpdatedAt)
                && status.equals(other.status)
                && remoteWasDeleted.equals(other.remoteWasDeleted)
                && createdAt.equals(other.createdAt)
                && modifiedAt.equals(other.modifiedAt)
                && fieldMappings.equals(other.fieldMappings)
                && remoteData.equals(other.remoteData);
    }

    @Override
    public int hashCode() {
        return Objects.hash(
                this.id,
                this.remoteId,
                this.application,
                this.jobInterviewStage,
                this.organizer,
                this.interviewers,
                this.location,
                this.startAt,
                this.endAt,
                this.remoteCreatedAt,
                this.remoteUpdatedAt,
                this.status,
                this.remoteWasDeleted,
                this.createdAt,
                this.modifiedAt,
                this.fieldMappings,
                this.remoteData);
    }

    @Override
    public String toString() {
        return ObjectMappers.stringify(this);
    }

    public static Builder builder() {
        return new Builder();
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static final class Builder {
        private Optional<String> id = Optional.empty();

        private Optional<String> remoteId = Optional.empty();

        private Optional<ScheduledInterviewApplication> application = Optional.empty();

        private Optional<ScheduledInterviewJobInterviewStage> jobInterviewStage = Optional.empty();

        private Optional<ScheduledInterviewOrganizer> organizer = Optional.empty();

        private Optional<List<Optional<ScheduledInterviewInterviewersItem>>> interviewers = Optional.empty();

        private Optional<String> location = Optional.empty();

        private Optional<OffsetDateTime> startAt = Optional.empty();

        private Optional<OffsetDateTime> endAt = Optional.empty();

        private Optional<OffsetDateTime> remoteCreatedAt = Optional.empty();

        private Optional<OffsetDateTime> remoteUpdatedAt = Optional.empty();

        private Optional<ScheduledInterviewStatus> status = Optional.empty();

        private Optional<Boolean> remoteWasDeleted = Optional.empty();

        private Optional<OffsetDateTime> createdAt = Optional.empty();

        private Optional<OffsetDateTime> modifiedAt = Optional.empty();

        private Optional<Map<String, JsonNode>> fieldMappings = Optional.empty();

        private Optional<List<RemoteData>> remoteData = Optional.empty();

        private Builder() {}

        public Builder from(ScheduledInterview other) {
            id(other.getId());
            remoteId(other.getRemoteId());
            application(other.getApplication());
            jobInterviewStage(other.getJobInterviewStage());
            organizer(other.getOrganizer());
            interviewers(other.getInterviewers());
            location(other.getLocation());
            startAt(other.getStartAt());
            endAt(other.getEndAt());
            remoteCreatedAt(other.getRemoteCreatedAt());
            remoteUpdatedAt(other.getRemoteUpdatedAt());
            status(other.getStatus());
            remoteWasDeleted(other.getRemoteWasDeleted());
            createdAt(other.getCreatedAt());
            modifiedAt(other.getModifiedAt());
            fieldMappings(other.getFieldMappings());
            remoteData(other.getRemoteData());
            return this;
        }

        @JsonSetter(value = "id", nulls = Nulls.SKIP)
        public Builder id(Optional<String> id) {
            this.id = id;
            return this;
        }

        public Builder id(String id) {
            this.id = Optional.of(id);
            return this;
        }

        @JsonSetter(value = "remote_id", nulls = Nulls.SKIP)
        public Builder remoteId(Optional<String> remoteId) {
            this.remoteId = remoteId;
            return this;
        }

        public Builder remoteId(String remoteId) {
            this.remoteId = Optional.of(remoteId);
            return this;
        }

        @JsonSetter(value = "application", nulls = Nulls.SKIP)
        public Builder application(Optional<ScheduledInterviewApplication> application) {
            this.application = application;
            return this;
        }

        public Builder application(ScheduledInterviewApplication application) {
            this.application = Optional.of(application);
            return this;
        }

        @JsonSetter(value = "job_interview_stage", nulls = Nulls.SKIP)
        public Builder jobInterviewStage(Optional<ScheduledInterviewJobInterviewStage> jobInterviewStage) {
            this.jobInterviewStage = jobInterviewStage;
            return this;
        }

        public Builder jobInterviewStage(ScheduledInterviewJobInterviewStage jobInterviewStage) {
            this.jobInterviewStage = Optional.of(jobInterviewStage);
            return this;
        }

        @JsonSetter(value = "organizer", nulls = Nulls.SKIP)
        public Builder organizer(Optional<ScheduledInterviewOrganizer> organizer) {
            this.organizer = organizer;
            return this;
        }

        public Builder organizer(ScheduledInterviewOrganizer organizer) {
            this.organizer = Optional.of(organizer);
            return this;
        }

        @JsonSetter(value = "interviewers", nulls = Nulls.SKIP)
        public Builder interviewers(Optional<List<Optional<ScheduledInterviewInterviewersItem>>> interviewers) {
            this.interviewers = interviewers;
            return this;
        }

        public Builder interviewers(List<Optional<ScheduledInterviewInterviewersItem>> interviewers) {
            this.interviewers = Optional.of(interviewers);
            return this;
        }

        @JsonSetter(value = "location", nulls = Nulls.SKIP)
        public Builder location(Optional<String> location) {
            this.location = location;
            return this;
        }

        public Builder location(String location) {
            this.location = Optional.of(location);
            return this;
        }

        @JsonSetter(value = "start_at", nulls = Nulls.SKIP)
        public Builder startAt(Optional<OffsetDateTime> startAt) {
            this.startAt = startAt;
            return this;
        }

        public Builder startAt(OffsetDateTime startAt) {
            this.startAt = Optional.of(startAt);
            return this;
        }

        @JsonSetter(value = "end_at", nulls = Nulls.SKIP)
        public Builder endAt(Optional<OffsetDateTime> endAt) {
            this.endAt = endAt;
            return this;
        }

        public Builder endAt(OffsetDateTime endAt) {
            this.endAt = Optional.of(endAt);
            return this;
        }

        @JsonSetter(value = "remote_created_at", nulls = Nulls.SKIP)
        public Builder remoteCreatedAt(Optional<OffsetDateTime> remoteCreatedAt) {
            this.remoteCreatedAt = remoteCreatedAt;
            return this;
        }

        public Builder remoteCreatedAt(OffsetDateTime remoteCreatedAt) {
            this.remoteCreatedAt = Optional.of(remoteCreatedAt);
            return this;
        }

        @JsonSetter(value = "remote_updated_at", nulls = Nulls.SKIP)
        public Builder remoteUpdatedAt(Optional<OffsetDateTime> remoteUpdatedAt) {
            this.remoteUpdatedAt = remoteUpdatedAt;
            return this;
        }

        public Builder remoteUpdatedAt(OffsetDateTime remoteUpdatedAt) {
            this.remoteUpdatedAt = Optional.of(remoteUpdatedAt);
            return this;
        }

        @JsonSetter(value = "status", nulls = Nulls.SKIP)
        public Builder status(Optional<ScheduledInterviewStatus> status) {
            this.status = status;
            return this;
        }

        public Builder status(ScheduledInterviewStatus status) {
            this.status = Optional.of(status);
            return this;
        }

        @JsonSetter(value = "remote_was_deleted", nulls = Nulls.SKIP)
        public Builder remoteWasDeleted(Optional<Boolean> remoteWasDeleted) {
            this.remoteWasDeleted = remoteWasDeleted;
            return this;
        }

        public Builder remoteWasDeleted(Boolean remoteWasDeleted) {
            this.remoteWasDeleted = Optional.of(remoteWasDeleted);
            return this;
        }

        @JsonSetter(value = "created_at", nulls = Nulls.SKIP)
        public Builder createdAt(Optional<OffsetDateTime> createdAt) {
            this.createdAt = createdAt;
            return this;
        }

        public Builder createdAt(OffsetDateTime createdAt) {
            this.createdAt = Optional.of(createdAt);
            return this;
        }

        @JsonSetter(value = "modified_at", nulls = Nulls.SKIP)
        public Builder modifiedAt(Optional<OffsetDateTime> modifiedAt) {
            this.modifiedAt = modifiedAt;
            return this;
        }

        public Builder modifiedAt(OffsetDateTime modifiedAt) {
            this.modifiedAt = Optional.of(modifiedAt);
            return this;
        }

        @JsonSetter(value = "field_mappings", nulls = Nulls.SKIP)
        public Builder fieldMappings(Optional<Map<String, JsonNode>> fieldMappings) {
            this.fieldMappings = fieldMappings;
            return this;
        }

        public Builder fieldMappings(Map<String, JsonNode> fieldMappings) {
            this.fieldMappings = Optional.of(fieldMappings);
            return this;
        }

        @JsonSetter(value = "remote_data", nulls = Nulls.SKIP)
        public Builder remoteData(Optional<List<RemoteData>> remoteData) {
            this.remoteData = remoteData;
            return this;
        }

        public Builder remoteData(List<RemoteData> remoteData) {
            this.remoteData = Optional.of(remoteData);
            return this;
        }

        public ScheduledInterview build() {
            return new ScheduledInterview(
                    id,
                    remoteId,
                    application,
                    jobInterviewStage,
                    organizer,
                    interviewers,
                    location,
                    startAt,
                    endAt,
                    remoteCreatedAt,
                    remoteUpdatedAt,
                    status,
                    remoteWasDeleted,
                    createdAt,
                    modifiedAt,
                    fieldMappings,
                    remoteData);
        }
    }
}

File: merge-java-core/src/main/kotlin/dev/merge/api/models/ats/ScheduledInterview.kt
package dev.merge.api.models.ats

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import dev.merge.api.core.ExcludeMissing
import dev.merge.api.core.JsonField
import dev.merge.api.core.JsonMissing
import dev.merge.api.core.JsonValue
import dev.merge.api.core.NoAutoDetect
import dev.merge.api.core.toUnmodifiable
import dev.merge.api.errors.MergeInvalidDataException
import dev.merge.api.models.*
import java.time.OffsetDateTime
import java.util.Objects
import java.util.Optional

/**
 * # The ScheduledInterview Object
 *
 * ### Description
 *
 * The `ScheduledInterview` object is used to represent a scheduled interview for a given
 * candidate’s application to a job. An `Application` can have multiple `ScheduledInterview`s
 * depending on the particular hiring process.
 *
 * ### Usage Example
 *
 * Fetch from the `LIST ScheduledInterviews` endpoint and filter by `interviewers` to show all
 * office locations.
 */
@JsonDeserialize(builder = ScheduledInterview.Builder::class)
@NoAutoDetect
class ScheduledInterview
private constructor(
    private val id: JsonField<String>,
    private val remoteId: JsonField<String>,
    private val application: JsonField<String>,
    private val jobInterviewStage: JsonField<String>,
    private val organizer: JsonField<String>,
    private val interviewers: JsonField<List<String>>,
    private val location: JsonField<String>,
    private val startAt: JsonField<OffsetDateTime>,
    private val endAt: JsonField<OffsetDateTime>,
    private val remoteCreatedAt: JsonField<OffsetDateTime>,
    private val remoteUpdatedAt: JsonField<OffsetDateTime>,
    private val status: JsonField<ScheduledInterviewStatusEnum>,
    private val remoteWasDeleted: JsonField<Boolean>,
    private val fieldMappings: JsonValue,
    private val modifiedAt: JsonField<OffsetDateTime>,
    private val remoteData: JsonField<List<RemoteData>>,
    private val additionalProperties: Map<String, JsonValue>,
) {

    private var validated: Boolean = false

    private var hashCode: Int = 0

    fun id(): Optional<String> = Optional.ofNullable(id.getNullable("id"))

    /** The third-party API ID of the matching object. */
    fun remoteId(): Optional<String> = Optional.ofNullable(remoteId.getNullable("remote_id"))

    /** The application being interviewed. */
    fun application(): Optional<String> =
        Optional.ofNullable(application.getNullable("application"))

    /** The stage of the interview. */
    fun jobInterviewStage(): Optional<String> =
        Optional.ofNullable(jobInterviewStage.getNullable("job_interview_stage"))

    /** The user organizing the interview. */
    fun organizer(): Optional<String> = Optional.ofNullable(organizer.getNullable("organizer"))

    /** Array of `RemoteUser` IDs. */
    fun interviewers(): Optional<List<String>> =
        Optional.ofNullable(interviewers.getNullable("interviewers"))

    /** The interview's location. */
    fun location(): Optional<String> = Optional.ofNullable(location.getNullable("location"))

    /** When the interview was started. */
    fun startAt(): Optional<OffsetDateTime> = Optional.ofNullable(startAt.getNullable("start_at"))

    /** When the interview was ended. */
    fun endAt(): Optional<OffsetDateTime> = Optional.ofNullable(endAt.getNullable("end_at"))

    /** When the third party's interview was created. */
    fun remoteCreatedAt(): Optional<OffsetDateTime> =
        Optional.ofNullable(remoteCreatedAt.getNullable("remote_created_at"))

    /** When the third party's interview was updated. */
    fun remoteUpdatedAt(): Optional<OffsetDateTime> =
        Optional.ofNullable(remoteUpdatedAt.getNullable("remote_updated_at"))

    /**
     * - `SCHEDULED` - SCHEDULED
     * - `AWAITING_FEEDBACK` - AWAITING_FEEDBACK
     * - `COMPLETE` - COMPLETE
     */
    fun status(): Optional<ScheduledInterviewStatusEnum> =
        Optional.ofNullable(status.getNullable("status"))

    /** Indicates whether or not this object has been deleted by third party webhooks. */
    fun remoteWasDeleted(): Optional<Boolean> =
        Optional.ofNullable(remoteWasDeleted.getNullable("remote_was_deleted"))

    /** This is the datetime that this object was last updated by Merge */
    fun modifiedAt(): Optional<OffsetDateTime> =
        Optional.ofNullable(modifiedAt.getNullable("modified_at"))

    fun remoteData(): Optional<List<RemoteData>> =
        Optional.ofNullable(remoteData.getNullable("remote_data"))

    @JsonProperty("id") @ExcludeMissing fun _id() = id

    /** The third-party API ID of the matching object. */
    @JsonProperty("remote_id") @ExcludeMissing fun _remoteId() = remoteId

    /** The application being interviewed. */
    @JsonProperty("application") @ExcludeMissing fun _application() = application

    /** The stage of the interview. */
    @JsonProperty("job_interview_stage")
    @ExcludeMissing
    fun _jobInterviewStage() = jobInterviewStage

    /** The user organizing the interview. */
    @JsonProperty("organizer") @ExcludeMissing fun _organizer() = organizer

    /** Array of `RemoteUser` IDs. */
    @JsonProperty("interviewers") @ExcludeMissing fun _interviewers() = interviewers

    /** The interview's location. */
    @JsonProperty("location") @ExcludeMissing fun _location() = location

    /** When the interview was started. */
    @JsonProperty("start_at") @ExcludeMissing fun _startAt() = startAt

    /** When the interview was ended. */
    @JsonProperty("end_at") @ExcludeMissing fun _endAt() = endAt

    /** When the third party's interview was created. */
    @JsonProperty("remote_created_at") @ExcludeMissing fun _remoteCreatedAt() = remoteCreatedAt

    /** When the third party's interview was updated. */
    @JsonProperty("remote_updated_at") @ExcludeMissing fun _remoteUpdatedAt() = remoteUpdatedAt

    /**
     * - `SCHEDULED` - SCHEDULED
     * - `AWAITING_FEEDBACK` - AWAITING_FEEDBACK
     * - `COMPLETE` - COMPLETE
     */
    @JsonProperty("status") @ExcludeMissing fun _status() = status

    /** Indicates whether or not this object has been deleted by third party webhooks. */
    @JsonProperty("remote_was_deleted") @ExcludeMissing fun _remoteWasDeleted() = remoteWasDeleted

    @JsonProperty("field_mappings") @ExcludeMissing fun _fieldMappings() = fieldMappings

    /** This is the datetime that this object was last updated by Merge */
    @JsonProperty("modified_at") @ExcludeMissing fun _modifiedAt() = modifiedAt

    @JsonProperty("remote_data") @ExcludeMissing fun _remoteData() = remoteData

    @JsonAnyGetter
    @ExcludeMissing
    fun _additionalProperties(): Map<String, JsonValue> = additionalProperties

    fun validate(): ScheduledInterview = apply {
        if (!validated) {
            id()
            remoteId()
            application()
            jobInterviewStage()
            organizer()
            interviewers()
            location()
            startAt()
            endAt()
            remoteCreatedAt()
            remoteUpdatedAt()
            status()
            remoteWasDeleted()
            modifiedAt()
            remoteData().map { it.forEach { it.validate() } }
            validated = true
        }
    }

    fun toBuilder() = Builder().from(this)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        return other is ScheduledInterview &&
            this.id == other.id &&
            this.remoteId == other.remoteId &&
            this.application == other.application &&
            this.jobInterviewStage == other.jobInterviewStage &&
            this.organizer == other.organizer &&
            this.interviewers == other.interviewers &&
            this.location == other.location &&
            this.startAt == other.startAt &&
            this.endAt == other.endAt &&
            this.remoteCreatedAt == other.remoteCreatedAt &&
            this.remoteUpdatedAt == other.remoteUpdatedAt &&
            this.status == other.status &&
            this.remoteWasDeleted == other.remoteWasDeleted &&
            this.fieldMappings == other.fieldMappings &&
            this.modifiedAt == other.modifiedAt &&
            this.remoteData == other.remoteData &&
            this.additionalProperties == other.additionalProperties
    }

    override fun hashCode(): Int {
        if (hashCode == 0) {
            hashCode =
                Objects.hash(
                    id,
                    remoteId,
                    application,
                    jobInterviewStage,
                    organizer,
                    interviewers,
                    location,
                    startAt,
                    endAt,
                    remoteCreatedAt,
                    remoteUpdatedAt,
                    status,
                    remoteWasDeleted,
                    fieldMappings,
                    modifiedAt,
                    remoteData,
                    additionalProperties,
                )
        }
        return hashCode
    }

    override fun toString() =
        "ScheduledInterview{id=$id, remoteId=$remoteId, application=$application, jobInterviewStage=$jobInterviewStage, organizer=$organizer, interviewers=$interviewers, location=$location, startAt=$startAt, endAt=$endAt, remoteCreatedAt=$remoteCreatedAt, remoteUpdatedAt=$remoteUpdatedAt, status=$status, remoteWasDeleted=$remoteWasDeleted, fieldMappings=$fieldMappings, modifiedAt=$modifiedAt, remoteData=$remoteData, additionalProperties=$additionalProperties}"

    companion object {

        @JvmStatic fun builder() = Builder()
    }

    class Builder {

        private var id: JsonField<String> = JsonMissing.of()
        private var remoteId: JsonField<String> = JsonMissing.of()
        private var application: JsonField<String> = JsonMissing.of()
        private var jobInterviewStage: JsonField<String> = JsonMissing.of()
        private var organizer: JsonField<String> = JsonMissing.of()
        private var interviewers: JsonField<List<String>> = JsonMissing.of()
        private var location: JsonField<String> = JsonMissing.of()
        private var startAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var endAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var remoteCreatedAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var remoteUpdatedAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var status: JsonField<ScheduledInterviewStatusEnum> = JsonMissing.of()
        private var remoteWasDeleted: JsonField<Boolean> = JsonMissing.of()
        private var fieldMappings: JsonValue = JsonMissing.of()
        private var modifiedAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var remoteData: JsonField<List<RemoteData>> = JsonMissing.of()
        private var additionalProperties: MutableMap<String, JsonValue> = mutableMapOf()

        @JvmSynthetic
        internal fun from(scheduledInterview: ScheduledInterview) = apply {
            this.id = scheduledInterview.id
            this.remoteId = scheduledInterview.remoteId
            this.application = scheduledInterview.application
            this.jobInterviewStage = scheduledInterview.jobInterviewStage
            this.organizer = scheduledInterview.organizer
            this.interviewers = scheduledInterview.interviewers
            this.location = scheduledInterview.location
            this.startAt = scheduledInterview.startAt
            this.endAt = scheduledInterview.endAt
            this.remoteCreatedAt = scheduledInterview.remoteCreatedAt
            this.remoteUpdatedAt = scheduledInterview.remoteUpdatedAt
            this.status = scheduledInterview.status
            this.remoteWasDeleted = scheduledInterview.remoteWasDeleted
            this.fieldMappings = scheduledInterview.fieldMappings
            this.modifiedAt = scheduledInterview.modifiedAt
            this.remoteData = scheduledInterview.remoteData
            additionalProperties(scheduledInterview.additionalProperties)
        }

        fun id(id: String) = id(JsonField.of(id))

        @JsonProperty("id") @ExcludeMissing fun id(id: JsonField<String>) = apply { this.id = id }

        /** The third-party API ID of the matching object. */
        fun remoteId(remoteId: String) = remoteId(JsonField.of(remoteId))

        /** The third-party API ID of the matching object. */
        @JsonProperty("remote_id")
        @ExcludeMissing
        fun remoteId(remoteId: JsonField<String>) = apply { this.remoteId = remoteId }

        /** The application being interviewed. */
        fun application(application: String) = application(JsonField.of(application))

        /** The application being interviewed. */
        @JsonProperty("application")
        @ExcludeMissing
        fun application(application: JsonField<String>) = apply { this.application = application }

        /** The stage of the interview. */
        fun jobInterviewStage(jobInterviewStage: String) =
            jobInterviewStage(JsonField.of(jobInterviewStage))

        /** The stage of the interview. */
        @JsonProperty("job_interview_stage")
        @ExcludeMissing
        fun jobInterviewStage(jobInterviewStage: JsonField<String>) = apply {
            this.jobInterviewStage = jobInterviewStage
        }

        /** The user organizing the interview. */
        fun organizer(organizer: String) = organizer(JsonField.of(organizer))

        /** The user organizing the interview. */
        @JsonProperty("organizer")
        @ExcludeMissing
        fun organizer(organizer: JsonField<String>) = apply { this.organizer = organizer }

        /** Array of `RemoteUser` IDs. */
        fun interviewers(interviewers: List<String>) = interviewers(JsonField.of(interviewers))

        /** Array of `RemoteUser` IDs. */
        @JsonProperty("interviewers")
        @ExcludeMissing
        fun interviewers(interviewers: JsonField<List<String>>) = apply {
            this.interviewers = interviewers
        }

        /** The interview's location. */
        fun location(location: String) = location(JsonField.of(location))

        /** The interview's location. */
        @JsonProperty("location")
        @ExcludeMissing
        fun location(location: JsonField<String>) = apply { this.location = location }

        /** When the interview was started. */
        fun startAt(startAt: OffsetDateTime) = startAt(JsonField.of(startAt))

        /** When the interview was started. */
        @JsonProperty("start_at")
        @ExcludeMissing
        fun startAt(startAt: JsonField<OffsetDateTime>) = apply { this.startAt = startAt }

        /** When the interview was ended. */
        fun endAt(endAt: OffsetDateTime) = endAt(JsonField.of(endAt))

        /** When the interview was ended. */
        @JsonProperty("end_at")
        @ExcludeMissing
        fun endAt(endAt: JsonField<OffsetDateTime>) = apply { this.endAt = endAt }

        /** When the third party's interview was created. */
        fun remoteCreatedAt(remoteCreatedAt: OffsetDateTime) =
            remoteCreatedAt(JsonField.of(remoteCreatedAt))

        /** When the third party's interview was created. */
        @JsonProperty("remote_created_at")
        @ExcludeMissing
        fun remoteCreatedAt(remoteCreatedAt: JsonField<OffsetDateTime>) = apply {
            this.remoteCreatedAt = remoteCreatedAt
        }

        /** When the third party's interview was updated. */
        fun remoteUpdatedAt(remoteUpdatedAt: OffsetDateTime) =
            remoteUpdatedAt(JsonField.of(remoteUpdatedAt))

        /** When the third party's interview was updated. */
        @JsonProperty("remote_updated_at")
        @ExcludeMissing
        fun remoteUpdatedAt(remoteUpdatedAt: JsonField<OffsetDateTime>) = apply {
            this.remoteUpdatedAt = remoteUpdatedAt
        }

        /**
         * - `SCHEDULED` - SCHEDULED
         * - `AWAITING_FEEDBACK` - AWAITING_FEEDBACK
         * - `COMPLETE` - COMPLETE
         */
        fun status(status: ScheduledInterviewStatusEnum) = status(JsonField.of(status))

        /**
         * - `SCHEDULED` - SCHEDULED
         * - `AWAITING_FEEDBACK` - AWAITING_FEEDBACK
         * - `COMPLETE` - COMPLETE
         */
        @JsonProperty("status")
        @ExcludeMissing
        fun status(status: JsonField<ScheduledInterviewStatusEnum>) = apply { this.status = status }

        /** Indicates whether or not this object has been deleted by third party webhooks. */
        fun remoteWasDeleted(remoteWasDeleted: Boolean) =
            remoteWasDeleted(JsonField.of(remoteWasDeleted))

        /** Indicates whether or not this object has been deleted by third party webhooks. */
        @JsonProperty("remote_was_deleted")
        @ExcludeMissing
        fun remoteWasDeleted(remoteWasDeleted: JsonField<Boolean>) = apply {
            this.remoteWasDeleted = remoteWasDeleted
        }

        @JsonProperty("field_mappings")
        @ExcludeMissing
        fun fieldMappings(fieldMappings: JsonValue) = apply { this.fieldMappings = fieldMappings }

        /** This is the datetime that this object was last updated by Merge */
        fun modifiedAt(modifiedAt: OffsetDateTime) = modifiedAt(JsonField.of(modifiedAt))

        /** This is the datetime that this object was last updated by Merge */
        @JsonProperty("modified_at")
        @ExcludeMissing
        fun modifiedAt(modifiedAt: JsonField<OffsetDateTime>) = apply {
            this.modifiedAt = modifiedAt
        }

        fun remoteData(remoteData: List<RemoteData>) = remoteData(JsonField.of(remoteData))

        @JsonProperty("remote_data")
        @ExcludeMissing
        fun remoteData(remoteData: JsonField<List<RemoteData>>) = apply {
            this.remoteData = remoteData
        }

        fun additionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.clear()
            this.additionalProperties.putAll(additionalProperties)
        }

        @JsonAnySetter
        fun putAdditionalProperty(key: String, value: JsonValue) = apply {
            this.additionalProperties.put(key, value)
        }

        fun putAllAdditionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.putAll(additionalProperties)
        }

        fun build(): ScheduledInterview =
            ScheduledInterview(
                id,
                remoteId,
                application,
                jobInterviewStage,
                organizer,
                interviewers.map { it.toUnmodifiable() },
                location,
                startAt,
                endAt,
                remoteCreatedAt,
                remoteUpdatedAt,
                status,
                remoteWasDeleted,
                fieldMappings,
                modifiedAt,
                remoteData.map { it.toUnmodifiable() },
                additionalProperties.toUnmodifiable(),
            )
    }

    class ScheduledInterviewStatusEnum
    @JsonCreator
    private constructor(
        private val value: JsonField<String>,
    ) {

        @com.fasterxml.jackson.annotation.JsonValue fun _value(): JsonField<String> = value

        override fun equals(other: Any?): Boolean {
            if (this === other) {
                return true
            }

            return other is ScheduledInterviewStatusEnum && this.value == other.value
        }

        override fun hashCode() = value.hashCode()

        override fun toString() = value.toString()

        companion object {

            @JvmField val SCHEDULED = ScheduledInterviewStatusEnum(JsonField.of("SCHEDULED"))

            @JvmField
            val AWAITING_FEEDBACK = ScheduledInterviewStatusEnum(JsonField.of("AWAITING_FEEDBACK"))

            @JvmField val COMPLETE = ScheduledInterviewStatusEnum(JsonField.of("COMPLETE"))

            @JvmStatic fun of(value: String) = ScheduledInterviewStatusEnum(JsonField.of(value))
        }

        enum class Known {
            SCHEDULED,
            AWAITING_FEEDBACK,
            COMPLETE,
        }

        enum class Value {
            SCHEDULED,
            AWAITING_FEEDBACK,
            COMPLETE,
            _UNKNOWN,
        }

        fun value(): Value =
            when (this) {
                SCHEDULED -> Value.SCHEDULED
                AWAITING_FEEDBACK -> Value.AWAITING_FEEDBACK
                COMPLETE -> Value.COMPLETE
                else -> Value._UNKNOWN
            }

        fun known(): Known =
            when (this) {
                SCHEDULED -> Known.SCHEDULED
                AWAITING_FEEDBACK -> Known.AWAITING_FEEDBACK
                COMPLETE -> Known.COMPLETE
                else ->
                    throw MergeInvalidDataException("Unknown ScheduledInterviewStatusEnum: $value")
            }

        fun asString(): String = _value().asStringOrThrow()
    }
}
